File: storage/serializer.py
"""JSON serialization for the canvas state.

The canvas is a collection of layers, each holding strokes (polylines of
pixel points with a colour and thickness).  This module converts that graph
to/from plain JSON-ready dictionaries so sessions can be autosaved and
restored.  The format is versioned (``version`` field) so future changes can
migrate old files instead of breaking them.

Serialization only touches the durable state (canvas size, layers, strokes,
active layer): undo/redo history is intentionally not persisted.
"""

from __future__ import annotations

from typing import Any

import config
from canvas.layers import Layer
from canvas.strokes import Stroke
from canvas.virtual_canvas import VirtualCanvas

CANVAS_FORMAT_VERSION = 1


def _stroke_to_dict(stroke: Stroke) -> dict[str, Any]:
    return {
        "points": [[x, y] for x, y in stroke.points],
        "color_hex": stroke.color_hex,
        "thickness": stroke.thickness,
        "layer_id": stroke.layer_id,
    }


def _layer_to_dict(layer: Layer) -> dict[str, Any]:
    return {
        "id": layer.id,
        "name": layer.name,
        "visible": bool(layer.visible),
        "opacity": float(layer.opacity),
        "strokes": [_stroke_to_dict(stroke) for stroke in layer.strokes],
    }


def canvas_to_dict(canvas: VirtualCanvas) -> dict[str, Any]:
    """Serialize a canvas (or any structural equivalent) to a JSON-ready dict."""
    return {
        "version": CANVAS_FORMAT_VERSION,
        "canvas": {
            "width": canvas.width,
            "height": canvas.height,
            "active_layer_id": canvas.active_layer.id,
        },
        "layers": [_layer_to_dict(layer) for layer in canvas.layers],
    }


def _stroke_from_dict(data: dict[str, Any]) -> Stroke:
    return Stroke(
        points=[(int(x), int(y)) for x, y in data["points"]],
        color_hex=data.get("color_hex", config.DEFAULT_COLOR),
        thickness=int(data.get("thickness", config.DEFAULT_BRUSH_SIZE)),
        layer_id=int(data.get("layer_id", 0)),
    )


def _layer_from_dict(data: dict[str, Any]) -> Layer:
    return Layer(
        id=int(data["id"]),
        name=data.get("name", "Layer"),
        visible=bool(data.get("visible", True)),
        opacity=float(data.get("opacity", 1.0)),
        strokes=[_stroke_from_dict(stroke) for stroke in data.get("strokes", [])],
    )


def canvas_from_dict(data: dict[str, Any]) -> VirtualCanvas:
    """Rebuild a :class:`VirtualCanvas` from :func:`canvas_to_dict` output.

    Raises ``ValueError`` for unsupported versions or malformed payloads.
    """
    if not isinstance(data, dict):
        raise ValueError("Autosave data is not an object")
    version = data.get("version")
    if version != CANVAS_FORMAT_VERSION:
        raise ValueError(f"Unsupported autosave format version: {version!r}")
    canvas_data = data.get("canvas")
    if not isinstance(canvas_data, dict):
        raise ValueError("Autosave data is missing the canvas section")
    layers = [_layer_from_dict(item) for item in data.get("layers", [])]
    if not layers:
        raise ValueError("Autosave data contains no layers")

    canvas = VirtualCanvas(
        width=int(canvas_data.get("width", config.CANVAS_WIDTH)),
        height=int(canvas_data.get("height", config.CANVAS_HEIGHT)),
    )
    canvas.layers = layers
    active_id = int(canvas_data.get("active_layer_id", layers[0].id))
    if not any(layer.id == active_id for layer in layers):
        active_id = layers[0].id
    canvas.set_active_layer(active_id)
    return canvas

File: packages/storage/src/serializer.ts
export const CANVAS_FORMAT_VERSION = 1;

export interface StrokeData {
  points: Array<{ x: number; y: number }>;
  colorHex: string;
  thickness: number;
  layerId: string;
}

export interface LayerData {
  id: string;
  name: string;
  visible: boolean;
  opacity: number;
  strokes: StrokeData[];
}

export interface CanvasSnapshot {
  width: number;
  height: number;
  backgroundColor: string;
  activeLayerId: string;
  layers: LayerData[];
}

function strokeToData(stroke: StrokeData): Record<string, unknown> {
  return {
    points: stroke.points.map((point) => [point.x, point.y]),
    colorHex: stroke.colorHex,
    thickness: stroke.thickness,
    layerId: stroke.layerId,
  };
}

function layerToData(layer: LayerData): Record<string, unknown> {
  return {
    id: layer.id,
    name: layer.name,
    visible: layer.visible,
    opacity: layer.opacity,
    strokes: layer.strokes.map(strokeToData),
  };
}

export function canvasToDict(canvas: CanvasSnapshot): Record<string, unknown> {
  return {
    version: CANVAS_FORMAT_VERSION,
    canvas: {
      width: canvas.width,
      height: canvas.height,
      backgroundColor: canvas.backgroundColor,
      activeLayerId: canvas.activeLayerId,
    },
    layers: canvas.layers.map(layerToData),
  };
}

function asRecord(value: unknown, message: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null) {
    throw new Error(message);
  }
  return value as Record<string, unknown>;
}

function strokeFromData(raw: unknown): StrokeData {
  const record = asRecord(raw, 'Malformed stroke data');
  const pointsRaw = Array.isArray(record.points) ? record.points : [];
  const points = pointsRaw.map((point) => {
    if (Array.isArray(point)) {
      const x = point[0];
      const y = point[1];
      return { x: x === undefined ? 0 : Number(x), y: y === undefined ? 0 : Number(y) };
    }
    const obj = asRecord(point, 'Malformed stroke point');
    return {
      x: obj.x === undefined ? 0 : Number(obj.x),
      y: obj.y === undefined ? 0 : Number(obj.y),
    };
  });
  return {
    points,
    colorHex: record.colorHex === undefined ? '#000000' : String(record.colorHex),
    thickness: record.thickness === undefined ? 4 : Number(record.thickness),
    layerId: record.layerId === undefined ? '0' : String(record.layerId),
  };
}

function layerFromData(raw: unknown): LayerData {
  const record = asRecord(raw, 'Malformed layer data');
  const strokesRaw = Array.isArray(record.strokes) ? record.strokes : [];
  return {
    id: String(record.id),
    name: record.name === undefined ? 'Layer' : String(record.name),
    visible: record.visible === undefined ? true : Boolean(record.visible),
    opacity: record.opacity === undefined ? 1 : Number(record.opacity),
    strokes: strokesRaw.map(strokeFromData),
  };
}

export function canvasFromDict(data: unknown): CanvasSnapshot {
  const root = asRecord(data, 'Autosave data is not an object');
  if (root.version !== CANVAS_FORMAT_VERSION) {
    throw new Error(`Unsupported autosave format version: ${String(root.version)}`);
  }
  const canvas = asRecord(root.canvas, 'Autosave data is missing the canvas section');
  const layersRaw = Array.isArray(root.layers) ? root.layers : [];
  if (layersRaw.length === 0) {
    throw new Error('Autosave data contains no layers');
  }
  const layers = layersRaw.map(layerFromData);
  const first = layers[0]!;
  let activeLayerId = canvas.activeLayerId === undefined ? first.id : String(canvas.activeLayerId);
  if (!layers.some((layer) => layer.id === activeLayerId)) {
    activeLayerId = first.id;
  }
  return {
    width: canvas.width === undefined ? 1280 : Number(canvas.width),
    height: canvas.height === undefined ? 720 : Number(canvas.height),
    backgroundColor: canvas.backgroundColor === undefined ? '#FFFFFF' : String(canvas.backgroundColor),
    activeLayerId,
    layers,
  };
}
